### PR TITLE
Added API for different usages

### PR DESCRIPTION
**Pull request by kaleanych**

### CODE REVIEW

File: ted_sws/core/entrypoints/api/main.py [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+@app.get("/fn/uuid/{value}")
+async def fn_uuid(
+        value: str,
+        input_process_type: Optional[UUIDInputProcessType] = UUIDInputProcessType.RAW,
+        ns_type: Optional[UUIDNamespaceType] = UUIDNamespaceType.DNS,
+        response_type: Optional[ResponseType] = ResponseType.RAW
+):
+    """
+    Returns generated UUID, taking as arguments:
+    - @value: string value to be hashed;
```
Comment (idellal) [inline ted_sws/core/entrypoints/api/main.py:115]: I don't see the cleaning step.
This function should has a Xpath as input to be cleaned up before using it to generate the UUID.
The cleaning process is explained here 
https://docs.google.com/document/d/1nnvD6XXYPSDzv_VukDHswzYKd_-PDUMe7E-kUGHNcc8/edit#bookmark=id.npqmufg93ynu

Comment (kaleanych) [reply ted_sws/core/entrypoints/api/main.py:115]: There is no clear statement on what algorithm should be used to clean the xpath

File: README.md [hunk 2]
```diff
@@ -211,8 +211,35 @@ Usage: sparql_runner [OPTIONS] [MAPPING_SUITE_ID]
 Options:
   -m, --opt-mappings-folder TEXT
   --help                                      Show this message and exit.
+```
 
+### API
+#### Start local API server
+To start the API server:
+```bash
+api-start-server
```
Comment (costezki) [inline README.md:220]: very well, however, will need to become a service started with Docker. We need to involve Dragos in deploying this. To get inspired on how docker files are written, you can look into any MDR project (rdf-validator-ws, rdf-fingerprinter-ws, rdf-differ-ws)

File: requirements.txt [hunk 3]
```diff
@@ -14,3 +14,5 @@ pymessagebus~=1.2
 click~=8.1.0
 openpyxl==3.0.9
 colorama~=0.4
+fastapi~=0.77
+uvicorn[standard]
```
Comment (costezki) [inline requirements.txt:18]: uvicorn, might have to be placed in another requirements.txt. 
I hope fresh environments are still installable with  pip, and for airflow.

File: ted_sws/core/entrypoints/api/main.py [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+
+import click
+from fastapi import FastAPI
+
+from ted_sws.event_manager.adapters.logger import Logger, LOG_INFO_LEVEL
+from ted_sws.event_manager.domain.message_bus import message_bus
+from ted_sws.event_manager.model.message import Log
+
+
+class ResponseType(Enum):
```
Comment (costezki) [inline ted_sws/core/entrypoints/api/main.py:16]: these shall be moved into a project component, called  `identity_manager` or something like that.

File: ted_sws/core/entrypoints/api/main.py [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+
+
+class UUIDNamespaceType(Enum):
+    DNS = "DNS"
+    URL = "URL"
+    OID = "OID"
+    X500 = "X500"
+
+
+API_SERVER_LOG_LEVELS: Tuple = ('critical', 'error', 'warning', 'info', 'debug', 'trace')
```
Comment (costezki) [inline ted_sws/core/entrypoints/api/main.py:33]: surely these constants are already available elsewhere and we do not need re-defining them here.

File: ted_sws/core/entrypoints/api/main.py [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+
+app = FastAPI()
+
+
+@click.command()
+@click.option('-h', '--host', default="localhost")
+@click.option('-p', '--port', default=8000, type=int)
+@click.option('-l', '--log-level', default="info", type=click.Choice(API_SERVER_LOG_LEVELS))
+def api_server_start(host, port, log_level):
```
Comment (costezki) [inline ted_sws/core/entrypoints/api/main.py:42]: this looks convenient for packaging into Docker.

File: ted_sws/core/entrypoints/api/main.py [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+    logger = Logger(name="API_SERVER", level=LOG_INFO_LEVEL)
+    fmt = "[%(asctime)s] - %(name)s - %(levelname)s:\n%(message)s"
+    date_fmt = "%Y-%m-%d %H:%M:%S"
+    logger_formatter = logging.Formatter(fmt, date_fmt)
+    logger.add_stdout_handler(formatter=logger_formatter)
+    log_level_text = f" --log-level {log_level}" if log_level else ""
+    bash_script = f"uvicorn --host {host} --port {port}{log_level_text} ted_sws.core.entrypoints.api.main:app --reload"
```
Comment (costezki) [inline ted_sws/core/entrypoints/api/main.py:49]: oh! surely there is a better way to do this.

Comment (costezki) [reply ted_sws/core/entrypoints/api/main.py:49]: we do not need to start a bash command, we can leave this aspect to the infra part. See how `gunicorn` is used in MDR projects.

File: ted_sws/core/entrypoints/api/main.py [hunk 8]
```diff
@@ -0,0 +1,124 @@
...
+    """
+    Generates MD5 hash, based on provided value
+    :param value:
+    :return:
+    """
+    return hashlib.md5(value.encode('utf8')).hexdigest()
+
+
+@app.get("/fn/md5/{value}")
+async def fn_md5(value: str, response_type: Optional[ResponseType] = ResponseType.RAW):
```
Comment (costezki) [inline ted_sws/core/entrypoints/api/main.py:80]: neat!

File: ted_sws/core/entrypoints/api/main.py [hunk 9]
```diff
@@ -0,0 +1,124 @@
...
+        uuid_ns = uuid.NAMESPACE_URL
+    elif ns_type == UUIDNamespaceType.OID:
+        uuid_ns = uuid.NAMESPACE_OID
+    elif ns_type == UUIDNamespaceType.X500:
+        uuid_ns = uuid.NAMESPACE_X500
+    return uuid_ns
+
+
+@app.get("/fn/uuid/{value}")
+async def fn_uuid(
```
Comment (costezki) [inline ted_sws/core/entrypoints/api/main.py:107]: sweet!

File: ted_sws/core/entrypoints/api/main.py (hunk 1 above)
Comment (costezki) [reply ted_sws/core/entrypoints/api/main.py:115]: We need additional functionality,as specified by Ibrahim, to perform also the cleanup, if needed (e.g. if a `clean` flag/parameter is `True`)

File: tests/unit/core/entrypoints/test_api.py [hunk 10]
```diff
@@ -0,0 +1,60 @@
+import uuid
+
+from click.testing import CliRunner
+from fastapi.testclient import TestClient
+
+from ted_sws.core.entrypoints.api.main import api_server_start as cli_api_server_start
+from ted_sws.core.entrypoints.api.main import app, single_result_response, string_md5, uuid_ns_by_type
+
+client = TestClient(app)
```
Comment (costezki) [inline tests/unit/core/entrypoints/test_api.py:9]: shouldn't this be a fixture?

File: tests/unit/core/entrypoints/test_api.py [hunk 11]
```diff
@@ -0,0 +1,60 @@
...
+cmdRunner = CliRunner()
```
Comment (costezki) [inline tests/unit/core/entrypoints/test_api.py:10]: shouldn't this be a fixture?

File: tests/unit/core/entrypoints/test_api.py [hunk 12]
```diff
@@ -0,0 +1,60 @@
...
+
+def test_api_fn_md5(input_value, response_type_json):
+    response = client.get(f"/fn/md5/{input_value}?response_type={response_type_json.value}")
+    assert response.status_code == 200
+    result_json = response.json()
+    assert "result" in result_json
+    assert result_json.get("result") != ""
+
+
+def test_api_fn_uuid(input_value, response_type_raw, uuid_input_process_type_md5):
```
Comment (costezki) [inline tests/unit/core/entrypoints/test_api.py:56]: let's add a few Negative tests, e.g. wrong parameters, or wrong path, for proper error handling.

**Review by costezki — APPROVED** [review]
Looks like a clean API implementation. 

There are a few ideas for improvement there. And some tasks for future deployment mechanisms.

File: ted_sws/core/entrypoints/api/main.py (hunk 7 above)
Comment (kaleanych) [reply ted_sws/core/entrypoints/api/main.py]: It is meant for local usage.

**Pull request merged**